src/grpc/stream_infer_handler.cc

```diff
@@ -130,362 +130,424 @@ ModelStreamInferHandler::StartNewRequest()
-        // state and enqueue itself to the completion queue to be
-        // taken up later. Otherwise, go to ISSUED state and wait
-        // for the callback to make a response available.
-        if (state->response_queue_->HasReadyResponse()) {
-          state->step_ = Steps::WRITEREADY;
-          state->context_->PutTaskBackToQueue(state);
-        } else {
-          state->step_ = Steps::ISSUED;
-        }
-      }
-    } else if (state->step_ == Steps::WRITEREADY) {
-      if (state->delay_response_ms_ != 0) {
-        // Will delay the write of the response by the specified time.
-        // This can be used to test the flow where there are other
-        // responses available to be written.
-        LOG_INFO << "Delaying the write of the response by "
-                 << state->delay_response_ms_ << " ms...";
-        std::this_thread::sleep_for(
-            std::chrono::milliseconds(state->delay_response_ms_));
-      }
-
-      // Finish the state if all the transactions associated with
-      // the state have completed.
-      if (state->IsComplete()) {
-        state->context_->DecrementRequestCounter();
-        finished = Finish(state);
-      } else {
-        // GRPC doesn't allow to issue another write till
-        // the notification from previous write has been
-        // delivered. If there is an ongoing write then
-        // defer writing and place the task at the back
-        // of the completion queue to be taken up later.
-        if (!state->context_->ongoing_write_) {
-          state->context_->ongoing_write_ = true;
-          state->context_->DecoupledWriteResponse(state);
-        } else {
-          state->context_->PutTaskBackToQueue(state);
-        }
-      }
+  // If the write failed (for example, client closed the stream)
+  // mark that the stream did not complete successfully but don't
+  // cancel right away... need to wait for any pending reads,
+  // inferences and writes to complete.
+  if (!rpc_ok) {
+    LOG_VERBOSE(1) << "Write for " << Name() << ", rpc_ok=" << rpc_ok
+                   << ", context " << state->context_->unique_id_ << ", "
+                   << state->unique_id_ << " step " << state->step_
+                   << ", failed";
+    state->context_->finish_ok_ = false;
+  }
+
+  // Log an error if 'state' is not the expected next response. Mark
+  // that the stream did not complete successfully but don't cancel
+  // right away... need to wait for any pending reads, inferences
+  // and writes to complete.
+  if (!state->context_->PopCompletedResponse(state)) {
+    LOG_ERROR << "Unexpected response for " << Name() << ", rpc_ok=" << rpc_ok
+              << ", context " << state->context_->unique_id_ << ", "
+              << state->unique_id_ << " step " << state->step_;
+    state->context_->finish_ok_ = false;
+  }
+
+  // Write the next response if it is ready...
+  state->context_->WriteResponseIfReady(nullptr);
```

Comment:
@tanmayv25 I don't understand this part. Do you know why we try to Write the next response (not the current' state's response) here in the WRITTEN state for a non-decoupled request? 

For the non-decoupled case, isn't every response's write handled when calling the [response callback](https://github.com/triton-inference-server/server/blob/6930d009b05a64a16eac3c751003c2329c429c6b/src/grpc/stream_infer_handler.cc#L633) or error handling for a given request?